### PR TITLE
Starting on implementing different variants

### PR DESCRIPTION
**Pull request by omarandlorraine**

### CODE REVIEW

**Comment by omarandlorraine** [comment]
From what I can see, the failing CI/CD process has to do with linting `skeptic`, a build time dependency. How then shall I proceed?

**Comment by mre** [comment]
Skeptic checks that the examples in the `README.md` are up-to-date. If these get fixed, then it should work again.

**Comment by omarandlorraine** [comment]
>Skeptic checks that the examples in the README.md are up-to-date. If these get fixed, then it should work again.

Can you read the output of the `CI / lint (pull request)`? The error is not coming from the README.

```
error: unnecessary hashes around raw string literal
  --> /home/runner/work/mos6502/mos6502/target/debug/build/mos6502-5d795f23ad8c7099/out/skeptic-tests.rs:53:10
   |
53 |   {}"####, r####"use mos6502::memory::Bus;
   |  __________^
54 | | use mos6502::memory::Memory;
55 | | use mos6502::instruction::Nmos6502;
56 | | use mos6502::cpu;
...  |
83 | | }
84 | | "####);
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_raw_string_hashes
   = note: `-D clippy::needless-raw-string-hashes` implied by `-D warnings`
```

**Comment by omarandlorraine** [comment]
I have disabled `--all-targets` for the time being, the tests are passing now

**Comment by mre** [comment]
Thanks. We can remove skeptic then. I hope there's a replacement of some sort.

**Comment by omarandlorraine** [comment]
> We can remove skeptic then. I hope there's a replacement of some sort.

Do you want to find one? Or shall we remove the example(s) from the README and refer readers to the `examples/` folder?

**Comment by omarandlorraine** [comment]
> Thanks. We can remove skeptic then. I hope there's a replacement of some sort.

There is a replacement; cargo doctest can do it now. I'll open another pull request to bring the compile times down.

**Comment by mre** [comment]
Perfect. Thanks!

**Comment by omarandlorraine** [comment]
> Perfect. Thanks!

You ready to merge the `variety` branch then, or? Still want to do the review?

**Comment by mre** [comment]
Oh, sorry, I was referring to your comment about removing skeptic and nom. 😅 
Still have to do the review of this PR.

**Comment by omarandlorraine** [comment]
I think we should revert [c87975e](https://github.com/mre/mos6502/pull/88/commits/c87975e9374e99b593ab20e6f87d7c01cc7ffd03) and then rebase this pull-request after #93 is in.

What do you reckon?

**Comment by mre** [comment]
> We should revert [c87975e](https://github.com/mre/mos6502/pull/88/commits/c87975e9374e99b593ab20e6f87d7c01cc7ffd03) and then rebase this pull-request after https://github.com/mre/mos6502/pull/93 is in

It's in now. Feel free to revert. 👍

**Comment by omarandlorraine** [comment]
> It's in now. Feel free to revert. 👍

That commit has been reverted and the PR has been rebased; so this is ready for merging as far as I am concerned

File: src/instruction.rs [hunk 1]
```diff
@@ -154,517 +156,307 @@ impl AddressingMode {
...
-    Some((Instruction::SBC, AddressingMode::AbsoluteX)),
-    /*0xFE*/
-    Some((Instruction::INC, AddressingMode::AbsoluteX)),
-    /*0xFF*/
-    None,
-];
+pub struct Nmos6502;
+
+impl crate::Variant for Nmos6502 {
+    fn decode(opcode: u8) -> Option<(Instruction, AddressingMode)> {
```
Comment (mre) [inline src/instruction.rs:162]: This part is so nice.

File: src/lib.rs [hunk 2]
```diff
@@ -40,3 +40,12 @@ pub mod cpu;
 pub mod instruction;
 pub mod memory;
 pub mod registers;
+
+pub trait Variant {
```
Comment (mre) [inline src/lib.rs:44]: Some doc comments would help here.

Comment (mre) [reply src/lib.rs:44]: Just a basic description that there are different variants of the CPU, and that they handle operations differently.

File: README.md [hunk 3]
```diff
@@ -52,7 +53,7 @@ fn main() {
         0x4c, 0x10, 0x00, // Jump to .algo
     ];
 
-    let mut cpu = cpu::CPU::new(Memory::new());
+    let mut cpu = cpu::CPU::new(Memory::new(), Nmos6502);
```
Comment (mre) [inline README.md:56]: Maybe we could add a `Default` impl for that?

```rust
// Initializes an Nmos6502 variant
let mut cpu = CPU::default();
```

If we don't want to add a default, we can alternatively provide

```rust
let mut cpu = CPU::nmos();
```

Just an idea.

**Comment by mre** [comment]
Looks good. Just some minor comments. We can merge if you like as there are no blockers. Up to you. 😄

**Comment by omarandlorraine** [comment]
Great, are you happy to merge and do a release on crates.io?

**Comment by mre** [comment]
Sure!

**Comment by mre** [comment]
Done: https://crates.io/crates/mos6502/versions

**Pull request merged**